ouroboros-consensus/src/ouroboros-consensus/Data/SOP/Match.hs

```diff
@@ -109,6 +116,53 @@ instance HAp (Mismatch f) where
       go (f :* _)  (MR fy gx) = MR fy (apFn f gx)
       go Nil       m          = case m of {}
 
+type instance Same (Mismatch f) = Mismatch f
+
+instance (forall x y. LiftedCoercible p p x y)
+      => HTrans (Mismatch p) (Mismatch p) where
+  htrans ::
+       forall proxy c f g xs ys. AllZipN (Prod (Mismatch p)) c xs ys
+    => proxy c
+    -> (forall x y. c x y => f x -> g y)
+    -> Mismatch p f xs
+    -> Mismatch p g ys
+  htrans p t = \case
+      ML fx gy -> ML (coerce fx) $ htrans p t gy
+      MR fy gx | Dict <- tailDict -> MR (hcoerce fy) $ t gx
+        where
+          tailDict :: Dict (AllZip (LiftedCoercible p p) (Tail xs) (Tail ys))
+          tailDict = replaceAllZipConstraint (Proxy @c)
+      MS m     -> MS $ htrans p t m
+
+  -- NOTE(jdral): this code could be replaced by 'unsafeCoerce' (see 'trans_NP'
+  -- or 'trans_NS' for examples), but this would technically sacrifice type
+  -- safety. For now, this version should be sufficient.
+  hcoerce ::
+       forall f g xs ys. AllZipN (Prod (Mismatch p)) (LiftedCoercible f g) xs ys
+    => Mismatch p f xs
+    -> Mismatch p g ys
+  hcoerce = \case
+      ML fx gy -> ML (coerce fx) $ hcoerce gy
+      MR fy gx | Dict <- tailDict -> MR (hcoerce fy) $ coerce gx
+        where
+          tailDict :: Dict (AllZip (LiftedCoercible p p) (Tail xs) (Tail ys))
+          tailDict = replaceAllZipConstraint (Proxy @(LiftedCoercible f g))
+      MS m     -> MS $ hcoerce m
```

Comment:
To avoid code duplication, we could also define
```haskell
defaultHcoerce ::
     forall h1 h2 f g xs ys.
     (HTrans h1 h2, AllZipN (Prod h1) (LiftedCoercible f g) xs ys)
  => h1 f xs
  -> h2 g ys
defaultHcoerce = htrans (Proxy @(LiftedCoercible f g)) coerce
```
and then write `hcoerce = defaultHcoerce` (also for `OptNP`).

Reply:
Good point -- I forgot about this

Reply:
Of course, we could also follow the `sop-core` convention and use `unsafeCoerce` instead. Do we want that?

Reply:
I have no intuition about how important/safe using `unsafeCoerce` here is.

Reply:
My intuition is that it `unsafeCoerce` is faster because it saves traversing the structure, but of course we should be reasonably sure that `unsafeCoerce` is "safe" within this context

Reply:
Let's write microbenchmarks for `hcoerce` :trollface: